formulaic/transforms/basis_spline.py

```diff
@@ -138,11 +145,18 @@ def basis_spline(  # pylint: disable=dangerous-default-value  # always replaced
                 raise ValueError(
                     f"Invalid value for `df`. `df` must be greater than {degree + (1 if include_intercept else 0)} [`degree` (+ 1 if `include_intercept` is `True`)]."
                 )
-            knots = list(
-                numpy.quantile(x, numpy.linspace(0, 1, nknots + 2))[1:-1].ravel()
+            if valid_x.sum(0) < nknots:
+                raise ValueError(
+                    f"Insufficient valid data points to compute {nknots} knots."
+                )
+            knots = cast(
+                List[float],
+                numpy.nanquantile(x[valid_x], numpy.linspace(0, 1, nknots + 2))[
```

Comment:
This truncation makes sense to me too. As you noted, this ensures that the clipped and nan extrapolations are consistent where they overlap. I think this was likely just an oversight in my original implementation; since it definitely doesn't make sense that you can have knots outside of your upper and lower bounds.